functions/go/set-project-id/main.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework/command"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+
+	kptv1 "github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/fix/v1"
+)
+
+var (
+	kccAPIVersionRegex = regexp.MustCompile(`^([\w]+)\.cnrm\.cloud\.google\.com\/[\w]+$`)
+)
+
+const (
+	projectIDSetterName = "project-id"
+	projectIDAnnotation = "cnrm.cloud.google.com/project-id"
+)
+
+func findSetterNode(nodes []*yaml.RNode, path string) (*yaml.RNode, error) {
+	for _, node := range nodes {
+		np := node.GetAnnotations()[kioutil.PathAnnotation]
+		if np == path {
+			return node, nil
+		}
+	}
+	return nil, fmt.Errorf(`file %s doesn't exist, please ensure the file specified in "configPath" exists and retry`, path)
+}
+
+func findKptfiles(nodes []*yaml.RNode) ([]*kptv1.KptFile, error) {
+	kptfiles := []*kptv1.KptFile{}
+	for _, node := range nodes {
+		if node.GetKind() == kptv1.KptFileName {
+			kf, err := kptv1.ReadFile(node)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read Kptfile: %w", err)
+			}
+			kptfiles = append(kptfiles, kf)
+		}
+	}
+	if len(kptfiles) == 0 {
+		return nil, fmt.Errorf("unable to find Kptfile, please include --include-meta-resources flag if a Kptfile is present")
+	}
+	return kptfiles, nil
+}
+
+func setKptfile(nodes []*yaml.RNode, kf *kptv1.KptFile) error {
+	b, err := yaml.Marshal(kf)
+	if err != nil {
+		return fmt.Errorf("failed to marshal updated Kptfile: %w", err)
+	}
+	kNode, err := yaml.Parse(string(b))
+	if err != nil {
+		return fmt.Errorf("failed to parse updated Kptfile: %w", err)
+	}
+
+	for i := range nodes {
+		if nodes[i].GetAnnotations()[kioutil.PathAnnotation] == kptv1.KptFileName {
```

Comment:
It seems there is a bug. This filter will always return the `Kptfile` from the root package.
I guess the filter should be something like:
```
 if nodes[i].GetAnnotations()[kioutil.PathAnnotation] == kf.GetAnnotations()[kioutil.PathAnnotation]
```

Please also add an e2e test for nesting packages.

Reply:
Done.